source/Calamari.Tests/KubernetesFixtures/HelmTemplateValueSourcesCreatorFixture.cs

```diff
@@ -0,0 +1,252 @@
+
+        [TestCaseSource(nameof(ParseTemplateValuesSourceTestData))]
+        public void ParseTemplateValuesSources_CorrectlyParsesAndWritesAndOrdersFiles(string templateValuesSourcesJson, IEnumerable<string> expectedFilenames, Dictionary<string, string> expectedFileContent, Dictionary<string, string> extraVariables = null)
+        {
+            // Arrange
+            var variables = new CalamariVariables
+            {
+                [SpecialVariables.Helm.TemplateValuesSources] = templateValuesSourcesJson,
+                [KnownVariables.OriginalPackageDirectoryPath] = RootDir
+            };
+            foreach (var kvp in extraVariables ?? new Dictionary<string, string>())
+            {
+                variables.Add(kvp.Key, kvp.Value);
+            }
+
+            var deployment = new RunningDeployment(variables)
+            {
+                CurrentDirectoryProvider = DeploymentWorkingDirectory.StagingDirectory
+            };
+
+            var fileSystem = Substitute.For<ICalamariFileSystem>();
+            fileSystem.EnumerateFilesWithGlob(Arg.Any<string>(), Arg.Any<string>())
+                      .Returns(ci => ci.ArgAt<string[]>(1)
+                                       ?.Select(x => Path.Combine(deployment.CurrentDirectory, x))
+                                       .ToArray());
+
+            var receivedFileContents = new Dictionary<string, string>();
+            fileSystem.When(fs => fs.WriteAllText(Arg.Any<string>(), Arg.Any<string>(), Arg.Any<Encoding>()))
+                      .Do(ci =>
+                          {
+                              receivedFileContents.Add(ci.ArgAt<string>(0), ci.ArgAt<string>(1));
+                          });
+
+            // Act
+            var filenames = HelmTemplateValueSourcesCreator.ParseTemplateValuesSources(deployment, fileSystem, new SilentLog());
+
+            // Assert
+            using (var _ = new AssertionScope())
+            {
+                filenames.Should().BeEquivalentTo(expectedFilenames.Select(f => Path.Combine(RootDir, f)));
+
+                foreach (var kvp in expectedFileContent)
+                {
+                    var filenameWithPath = Path.Combine(RootDir, kvp.Key);
+                    var expectedContent = kvp.Value;
+
+                    if (receivedFileContents.TryGetValue(filenameWithPath, out var receivedContent))
+                    {
+                        expectedContent.Should().Be(receivedContent);
+                        expectedContent.ToArray().Should().BeEquivalentTo(receivedContent.ToArray());
+                        Encoding.UTF8.GetBytes(expectedContent).Should().BeEquivalentTo(Encoding.UTF8.GetBytes(receivedContent));
+                    }
+
+                    fileSystem.Received(1).WriteAllText(Arg.Is(filenameWithPath), Arg.Is(expectedContent));
+                }
+            }
+        }
+
+        public static IEnumerable<object> ParseTemplateValuesSourceTestData => new object[]
```

Comment:
Related to my comments below, these test cases aren't particularly easy to grok, especially because we're really testing the functionality of multiple things here - the `GenerateXYZ` functions and the `HelmTemplateValueSourcesCreator` class.